File: shared/intervals.ts
export function unionIntervalDuration(
  intervals: Iterable<readonly [start: number, end: number]>
): number {
  const sorted = [...intervals]
    .filter(([start, end]) => end > start)
    .sort((left, right) => left[0] - right[0]);
  let total = 0;
  let currentStart: number | undefined;
  let currentEnd: number | undefined;
  for (const [start, end] of sorted) {
    if (currentStart === undefined || currentEnd === undefined) {
      currentStart = start;
      currentEnd = end;
    } else if (start <= currentEnd) {
      currentEnd = Math.max(currentEnd, end);
    } else {
      total += currentEnd - currentStart;
      currentStart = start;
      currentEnd = end;
    }
  }
  return currentStart === undefined || currentEnd === undefined
    ? total
    : total + currentEnd - currentStart;
}

File: backend/app/intervals.py
from __future__ import annotations

from collections.abc import Iterable


def union_interval_duration(intervals: Iterable[tuple[float, float]]) -> float:
    sorted_intervals = sorted(
        (start, end) for start, end in intervals if end > start
    )
    total = 0.0
    current_start: float | None = None
    current_end: float | None = None
    for start, end in sorted_intervals:
        if current_start is None or current_end is None:
            current_start, current_end = start, end
        elif start <= current_end:
            current_end = max(current_end, end)
        else:
            total += current_end - current_start
            current_start, current_end = start, end
    if current_start is not None and current_end is not None:
        total += current_end - current_start
    return total
